fe/fe-core/src/main/java/org/apache/doris/datasource/DataSourceMgr.java

```diff
@@ -52,103 +66,180 @@ public DataSourceMgr() {
 
     private void initInternalDataSource() {
         internalDataSource = new InternalDataSource();
-        idToDataSource.put(internalDataSource.getId(), internalDataSource);
         nameToDataSource.put(internalDataSource.getName(), internalDataSource);
     }
 
-    private void registerNewDataSource(ExternalDataSource ds) {
-        // TODO
-    }
-
     public InternalDataSource getInternalDataSource() {
         return internalDataSource;
     }
 
+    private void writeLock() {
+        lock.writeLock().lock();
+    }
+
+    private void writeUnlock() {
+        lock.writeLock().unlock();
+    }
+
+    private void readLock() {
+        lock.readLock().lock();
+    }
+
+    private void readUnlock() {
+        lock.readLock().unlock();
+    }
+
     /**
-     * get data source by id.
-     *
-     * @param id
-     * @param e
-     * @param <E>
-     * @return
-     * @throws E
+     * Create and hold the datasource instance and write the meta log.
      */
-    public <E extends MetaNotFoundException> DataSourceIf getDataSourceOrException(long id,
-            java.util.function.Function<Long, E> e) throws E {
-        DataSourceIf ds = idToDataSource.get(id);
-        if (ds == null) {
-            throw e.apply(id);
+    public void createDatasource(CreateCatalogStmt stmt) throws UserException {
```

Comment:
For `create/drop/alter` ds, your main logic is like:
```
DataSourceLog log = DataSourceFactory.xxx();
replayXXX(log);
Catalog.getCurrentCatalog().getEditLog().logDatasourceLog();
```

It works but may have some side effect.
Because in the `replayXXX` methods, there are some validation logic such as "check if ds exist", "check if property is valid".
And these `replayXXX` methods are also be called in replay editlog logic.
But in replay logic, there shouldn't be any exception thrown. So it is not suitable to put validation logic
in `replayXXX` methods.
`replayXXX` should only do the final metadata modification, other logic can be placed outside it.